### ISSUE TITLE
Idea for baseline

### ISSUE THREAD

**Issue by GolovanovSrg** [reinforceio/tensorforce#118]
For baseline usefully use body of main neural network and create layer for value function.

**Comment by michaelschaarschmidt** [comment]
Hi,
thanks for the suggestion, it's something we discussed on the weekend as we updated baselines for multi-state support.

The question is what we consider the body of the main neural network. In the case of DQN with convolutional layers + dense layers, just the convolutions? All of it? How can it be configured from the client side? 

We were still debating on how to do it, do you have any suggestions?

**Comment by GolovanovSrg** [comment]
You can do the same as internal_input and internal_output in the network builder - baseline_output. And by default as baseline_output  you can take network builder output.

**Comment by bigtreeliujie** [comment]
I understand @AlexKuhnle keeping network builder a clean feature. In my option,  this abstraction of network is a good idea and should be insisted during Tensorforce's whole life cycle.

In the other hand, sharing some network layers or even complicated network architectures between baseline network and policy network is also a common scene in RL. Using baseline means there is not only one network in the model. Even worse,  some new algorithm involves more networks. For example, , in [ICM](https://arxiv.org/abs/1705.05363) with AC there is four network: policy network, value network, forward model, inverse model, and some parts of those network are shared among them. 
I do not think the  "core network" is a good idea. Because there may be many "core network": network a and network b share a core, network b and c share another core ... 
As a fundamental library, Tensorfoce should provide users the flexibility to implemented their complicated networks, supporting sharing parts among networks.

Sharing network parts between networks is one of the features of Tensorflow. As a result, the network builder could use this feature in something like below:

```python
def network_part_builder(x, reuse=False):
    with tf.variable_scope(name="core1", reuse=reuse):
        x = tf_fnn(x, ...)
        ...
    return x

def policy_network_builder(x):
    core_output = network_part_builder(x)
    x = fnn(core_output, ...)
    return tf_softmax(x)

def value_network_builder(x):
    core_output = network_part_builder(x)
    return fnn(core_output, output=1, ...)
```

In current version baseline is implemented having its own predict and update interface, both interfaces do tf.session.run. In my option, this is not a good idea, because multiple session.run means several IO between python and Tensorflow. However, the prediction results of baseline getting form Tensorflow are immediately sent back to Tensorflow when update policy network.  
We could do the prediction when update the policy network in one Tensorflow run:
  - create baseline's network using the policy network' placeholders (Model.state, Model.reward) as inputs
  - in reward_estimation: do not predict baseline and do not sub discounted_rewards by baselines' predict value either.
  - in RLModel.create_tf_operations: ` self.loss_per_instance = -log_prob * (self.reward - tf.stop_gradient(baseline_network.output) `
We could also update baseline when update policy network:
  - put baseline network loss in tf.losses:  ` tf.losses.add_loss(baseline.loss) `
  - if baseline shares network parts with policy network, it seems good to add a coefficient parameter to the loss: ` tf.losses.add_loss(baseline.loss * baseline_coefficient) `

**Comment by AlexKuhnle** [comment]
Hey @bigtreeliujie, thanks for the detailed comments. The points you make in the end is what we're actually doing in this new soon-to-come "optimizer update", which really changes many more things than just optimizers. As you describe, the baseline will be part of the entire model TensorFlow graph instead of a separate session call. Only a minor difference to your description: We do not add the baseline loss to the model update, but rather add a baseline optimize operation to the model optimize operation. This allows to use a different optimizer for the baseline, which is what one probably wants to do particularly in the case of the natural gradient optimizer as used in TRPO.

Regarding the networks: Our *'philosophy'* usually was to keep everything as configurable as possible, unless it would mean giving up aspects of the high-level structure that we impose on all models (or at least the more specific sub-classes of `Model`). I think you can get pretty far with parameter sharing by using custom network definitions as you describe, but there might be limits. There will definitely be the option to define your own baseline externally, and there will probably be the option to either use the raw states as input or use the output of the model network as input. Definitely check out our *'beta version'* branch of the new library design, which we will put on GitHub by October, hopefully. :-)

**Comment by bigtreeliujie** [comment]
Thanks @AlexKuhnle , I am looking forward the beta release.

**Comment by michaelschaarschmidt** [comment]
This is addressed in dev via the network-baseline option

**Comment by lefnire** [comment]
Just double checking I understand the new setup. In `baseline_mode='network'` you have:

1. Main network
2. Baseline = Main + extra

So baseline is always "bigger" than main, yeah? Ideally we'd be able to _cut_ earlier (just share conv layers, then two separate deep MLP heads). Lacking that, my intuition for the current setup is it's saner to _not_ use `mode=network`, seems like it'd be difficult to fit. Instead use two separate nets (`baseline_mode='states', baseline: {type: 'mlp', sizes: [...]}`) - am I mistaken?

Anyway, I see how cutting early could be tough code-wise. In the interim I'm using a custom net for baseline, but wanted to custom-define the net. Had to do this https://github.com/lefnire/tensorforce/commit/e12b50a43520f9a7a626e5d200c12856e64571e4, LMK if there's an easier approach than that, or if not I can work towards a PR.

**Comment by AlexKuhnle** [comment]
You're right, the `network` mode currently is not ideal, since it leads to baseline being the 'larger' model. It seems like a two-part model might be the way to go. What we had in mind, and we wanted to add this in a smaller update soon-ish: Action-specific network parts extending the core network like the baseline does, so in principle one can control the size of baseline vs main model.

The way you integrate a custom baseline network is how it should work, essentially. However, your specific case of a generic `CustomBaseline` class is of course a bit weird. If you instead would add the `NetworkBaseline` to the dict of predefined baselines, that should be equivalent, right? That's probably the best solution, and I think was just forgotten...

**Comment by lefnire** [comment]
> If you instead would add the NetworkBaseline to the dict of predefined baselines, that should be equivalent, right?

Oh duh *facepalm*. I think it might be generally useful, I'll do up a PR.
